[docs] Updates installing expo updates with bare React (#19542)
expo/expo: docs/pages/bare/installing-updates.mdx

title: Installing expo-updates
---

import { DiffBlock } from '~/ui/components/Snippet';
import { Collapsible } from '~/ui/components/Collapsible';
import { Terminal } from '~/ui/components/Snippet/Blocks/Terminal';

> The expo-updates library fetches and manages updates received from a remote server. It supports [EAS Update](/eas-upadate/introduction), a hosted service that serves updates for projects using the expo-updates library.

Reviewer: ```suggestion
The `expo-updates` library fetches and manages updates received from a remote server. It supports [EAS Update](/eas-upadate/introduction), a hosted service that serves updates for projects using the expo-updates library.
```

We use inline syntax highlight for any library name or package name. also, the rest of the doc does use it for `expo-updates`. I think we should stick to it here too.

> If you are creating a new project, we recommend using `npx create-react-native-app` instead of `npx react-native init` because it will handle the following configuration for you automatically. It includes the expo-updates [config plugin](/guides/config-plugins.mdx), which will handle the following steps for you.

## Installation

> The expo-updates library requires that your project already has [Expo modules configured](/bare/installing-expo-modules.mdx). Be sure to install it before continuing.

Reviewer: ```suggestion
The `expo-updates` library requires that your project already has [Expo modules configured](/bare/installing-expo-modules.mdx). Be sure to install it before continuing.
```

> To get started, install expo-updates:

Reviewer: ```suggestion
To get started, install `expo-updates`:
```

<Terminal cmd={["$ npx expo install expo-updates"]} />

Then, install pods:

<Terminal cmd={["$ npx pod-install"]} />

> Once installation is complete, apply the changes from the following diffs to configure expo-updates in your project.

Reviewer: ```suggestion
Once installation is complete, apply the changes from the following diffs to configure `expo-updates` in your project.
```

## Configuration for Android

Apply the following configurations to your **AndroidManifest.xml** file:

```diff AndroidManifest.xml
+ <meta-data android:name="expo.modules.updates.EXPO_UPDATE_URL" android:value="https://exp.host/@my-expo-username/my-app" />
+ <meta-data android:name="expo.modules.updates.EXPO_SDK_VERSION" android:value="46.0.0" />
+ <meta-data android:name="expo.modules.updates.UPDATES_CONFIGURATION_REQUEST_HEADERS_KEY" android:value="{'expo-channel-name':'your-channel-name'}"/>
+ <meta-data android:name="expo.modules.updates.EXPO_UPDATES_ENABLED" android:value="true" />
+ <meta-data android:name="expo.modules.updates.EXPO_UPDATES_LAUNCH_WAIT_MS" android:value="0" />
+ <meta-data android:name="expo.modules.updates.EXPO_UPDATES_CHECK_ON_LAUNCH" android:value="ALWAYS" />
````

## Customizing Automatic Setup for iOS

> By default, `expo-updates` requires no additional setup. If you want to customize the installation, e.g. to enable updates only in some build variants, you can instead follow these manual setup steps and then apply any customizations.

Reviewer: ```suggestion
By default, `expo-updates` requires no additional setup. If you want to customize the installation, for example, to enable updates only in some build variants, you can instead follow these manual setup steps and then apply any customizations.
```

### Update AppDelegate.h

```diff apps/bare-update/ios/bareupdate/AppDelegate.h
#import <Foundation/Foundation.h>
+ #import <EXUpdates/EXUpdatesAppController.h>
#import <React/RCTBridgeDelegate.h>
#import <UIKit/UIKit.h>

#import <ExpoModulesCore/EXAppDelegateWrapper.h>

- @interface AppDelegate : EXAppDelegateWrapper <RCTBridgeDelegate>
+ @interface AppDelegate : EXAppDelegateWrapper <RCTBridgeDelegate, EXUpdatesAppControllerDelegate>
```

### Update AppDelegate.mm

> There are multiple changes to apply to your project's AppDelegate.mm.

Reviewer: ```suggestion
There are multiple changes to apply to your project's **AppDelegate.mm**.
```

Since it's a file name, we should make it bold [as per our styleguide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#filenames-as-bold-text).

-
-  return YES;
+  return bridge;
  }

 - (NSArray<id<RCTBridgeModule>> *)extraModulesForBridge:(RCTBridge *)bridge
```

```diff apps/bare-update/ios/bareupdate/AppDelegate.mm
  #ifdef DEBUG
   return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
  #else
-  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
+  return [[EXUpdatesAppController sharedInstance] launchAssetUrl];
  #endif
 }

+ - (void)appController:(EXUpdatesAppController *)appController didStartWithSuccess:(BOOL)success {
+  appController.bridge = [self initializeReactNativeApp];
+ }

 // Linking API
 - (BOOL)application:(UIApplication *)application openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
   return [RCTLinkingManager application:application openURL:url options:options];
```

### Expo.plist

```diff apps/bare-update/ios/bareupdate/Supporting/Expo.plist
     <key>EXUpdatesURL</key>
     <string>https://exp.host/@my-expo-username/my-app</string>
+    <key>EXUpdatesAutoSetup</key>
+    <false/>
   </dict>
 </plist>
```

## Customizing Automatic Setup for Android

> By default, `expo-updates` requires no additional setup. If you want to customize the installation, e.g. to enable updates only in some build variants, you can instead follow these manual setup steps and then apply any customizations.

Reviewer: ```suggestion
By default, `expo-updates` requires no additional setup. If you want to customize the installation, for example, to enable updates only in some build variants, you can instead follow these manual setup steps and then apply any customizations.
```

Unrelated to the changes in this PR but I think we can replace the latin abbreviation here as per our [styleguide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#abbreviations).

     }
+
+    @Override
+    protected @Nullable String getJSBundleFile() {
+      if (BuildConfig.DEBUG) {
+        return super.getJSBundleFile();
+      } else {
+        return UpdatesController.getInstance().getLaunchAssetFile();
+      }
+    }
+
+    @Override
+    protected @Nullable String getBundleAssetName() {
+      if (BuildConfig.DEBUG) {
+        return super.getBundleAssetName();
+      } else {
+        return UpdatesController.getInstance().getBundleAssetName();
+      }
+    }
   });
```

```diff apps/bare-update/android/app/src/main/java/com/bareupdate/MainApplication.java
     super.onCreate();
     SoLoader.init(this, /* native exopackage */ false);

+    if (!BuildConfig.DEBUG) {
+      UpdatesController.initialize(this);
+    }
+
     initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
     ApplicationLifecycleDispatcher.onApplicationCreate(this);
   }
```

## Usage

See more information about usage in the [expo-updates README](https://github.com/expo/expo/tree/main/packages/expo-updates/README.md).

## FAQ

<Collapsible summary="How do I customize which assets are included in an update bundle?">

> If you have assets (such as images or other media) that are imported in your application code, and you would like these to be downloaded atomically as part of an update, add the `assetBundlePatterns` field under the `expo` key in your project's app.json. This field should be an array of file glob strings which point to the assets you want bundled. For example: `"assetBundlePatterns": ["**/*"]`

Reviewer: ```suggestion
If you have assets (such as images or other media) that are imported in your application code, and you would like these to be downloaded atomically as part of an update, add the `assetBundlePatterns` field under the `expo` key in your project's **app.json**. This field should be an array of file glob strings that point to the assets you want to be bundled. For example: `"assetBundlePatterns": ["**/*"]`
```
